[WebGPU] Add native app for playing around with WebGPU

litherum opened the pull request:
#### d0dec94a73aab50bb7c5e50c391e3948676694c2
<pre>
[WebGPU] Add native app for playing around with WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=250626">https://bugs.webkit.org/show_bug.cgi?id=250626</a>
rdar://104265536

Reviewed by Tadeu Zagallo.

This patch creates a native multiplatform app (that has a GUI) that is hooked up to
WebGPU. The GUI is important so you can see the results of your drawing commands.
The app itself is implemented using SwiftUI and MetalKit, so it can be multiplatform
(both macOS and iOS). The reason to create this native app is mainly to ease build
times, so that it&apos;s possible to develop WebGPU without building all of WebKit, and because
it will be faster it iterate on because its compositing architecture is way simpler
than WebKit&apos;s.

There was already a command-line playground in WebGPU.xcodeproj; this patch moves it
to be alongside the new app playground inside Tools/.

WebGPU doesn&apos;t actually work with this native app out-of-the-box just yet; it needs
<a href="https://github.com/WebKit/WebKit/pull/8665.">https://github.com/WebKit/WebKit/pull/8665.</a>

* Source/WebGPU/CommandLinePlayground/main.swift: Removed.
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/WebGPUPlayground/AppPlayground/AppPlayground-Bridging-Header.h: Added.
* Tools/WebGPUPlayground/AppPlayground/AppPlayground.entitlements: Added.
* Tools/WebGPUPlayground/AppPlayground/AppPlaygroundApp.swift: Added.
(AppPlaygroundApp.body):
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/AccentColor.colorset/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/AppIcon.appiconset/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/ContentView.swift: Added.
(ContentView.body):
(ContentView_Previews.previews):
* Tools/WebGPUPlayground/AppPlayground/Preview Content/Preview Assets.xcassets/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.h: Added.
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m: Added.
(-[Scheduler init]):
(-[Scheduler schedule:]):
(-[WebGPUCoordinator init]):
(-[WebGPUCoordinator dealloc]):
(-[WebGPUCoordinator setView:]):
(-[WebGPUCoordinator drawInMTKView:]):
(-[WebGPUCoordinator mtkView:drawableSizeWillChange:]):
* Tools/WebGPUPlayground/AppPlayground/WebGPUView+AppKit.swift: Added.
(WebGPUView.makeNSView(_:)):
(WebGPUView.updateNSView(_:context:)):
(WebGPUView.makeCoordinator):
* Tools/WebGPUPlayground/AppPlayground/WebGPUView+UIKit.swift: Added.
(WebGPUView.makeUIView(_:)):
(WebGPUView.updateUIView(_:context:)):
(WebGPUView.makeCoordinator):
* Tools/WebGPUPlayground/AppPlayground/config.h: Added.
* Tools/WebGPUPlayground/CommandLinePlayground/main.c: Added.
(main):
* Tools/WebGPUPlayground/Configurations/AppPlayground.xcconfig: Copied from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/Configurations/Base.xcconfig: Added.
* Tools/WebGPUPlayground/Configurations/CommandLinePlayground.xcconfig: Copied from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/Configurations/DebugRelease.xcconfig: Renamed from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/WebGPUPlayground.xcodeproj/project.pbxproj: Added.

Canonical link: <a href="https://commits.webkit.org/258998@main">https://commits.webkit.org/258998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5973c380c3dccb5f14af92f43d8d12e545eb2b28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12698 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36536 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112818 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3596 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95836 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111985 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109355 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10573 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92397 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25242 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6076 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26643 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6257 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12238 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46154 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6179 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8009 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/AppPlayground/WebGPUView.swift:
```diff
@@ -0,0 +1,159 @@
+                return
+            }
+            if swapChain != nil {
+                wgpuSwapChainRelease(device)
```
Whoopsies

litherum commented on Source/WebGPU/AppPlayground/Utilities.m:
```diff
@@ -0,0 +1,33 @@
+void* cast(CAMetalLayer *layer)
+{
+    // I don't know how to do this in Swift, so I guess we can just do it here.
+    return (__bridge void *)(layer);
```
I guess this could just be `return (__bridge void *)layer;`

tadeuzagallo commented on Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m:
```diff
@@ -0,0 +1,229 @@
+
+- (void)schedule:(WGPUWorkItem)workItem
+{
+    [timers addObject:[NSTimer scheduledTimerWithTimeInterval:0 repeats:NO block:^(NSTimer *timer) {
```
I'm just curious why you used timers instead of GCD?

litherum replied:
Oh, it's just a bit more user-friendly.

tadeuzagallo commented on Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m:
```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2022 Apple Inc. All rights reserved.
```
Not sure how critical this is, but all your licenses still say 2022

litherum replied:
😳

litherum committed:
[WebGPU] Add native app for playing around with WebGPU
https://bugs.webkit.org/show_bug.cgi?id=250626
rdar://104265536

Reviewed by Tadeu Zagallo.

This patch creates a native multiplatform app (that has a GUI) that is hooked up to
WebGPU. The GUI is important so you can see the results of your drawing commands.
The app itself is implemented using SwiftUI and MetalKit, so it can be multiplatform
(both macOS and iOS). The reason to create this native app is mainly to ease build
times, so that it's possible to develop WebGPU without building all of WebKit, and because
it will be faster it iterate on because its compositing architecture is way simpler
than WebKit's.

There was already a command-line playground in WebGPU.xcodeproj; this patch moves it
to be alongside the new app playground inside Tools/.

WebGPU doesn't actually work with this native app out-of-the-box just yet; it needs
https://github.com/WebKit/WebKit/pull/8665.

* Source/WebGPU/CommandLinePlayground/main.swift: Removed.
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/WebGPUPlayground/AppPlayground/AppPlayground-Bridging-Header.h: Added.
* Tools/WebGPUPlayground/AppPlayground/AppPlayground.entitlements: Added.
* Tools/WebGPUPlayground/AppPlayground/AppPlaygroundApp.swift: Added.
(AppPlaygroundApp.body):
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/AccentColor.colorset/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/AppIcon.appiconset/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/Assets.xcassets/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/ContentView.swift: Added.
(ContentView.body):
(ContentView_Previews.previews):
* Tools/WebGPUPlayground/AppPlayground/Preview Content/Preview Assets.xcassets/Contents.json: Added.
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.h: Added.
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m: Added.
(-[Scheduler init]):
(-[Scheduler schedule:]):
(-[WebGPUCoordinator init]):
(-[WebGPUCoordinator dealloc]):
(-[WebGPUCoordinator setView:]):
(-[WebGPUCoordinator drawInMTKView:]):
(-[WebGPUCoordinator mtkView:drawableSizeWillChange:]):
* Tools/WebGPUPlayground/AppPlayground/WebGPUView+AppKit.swift: Added.
(WebGPUView.makeNSView(_:)):
(WebGPUView.updateNSView(_:context:)):
(WebGPUView.makeCoordinator):
* Tools/WebGPUPlayground/AppPlayground/WebGPUView+UIKit.swift: Added.
(WebGPUView.makeUIView(_:)):
(WebGPUView.updateUIView(_:context:)):
(WebGPUView.makeCoordinator):
* Tools/WebGPUPlayground/AppPlayground/config.h: Added.
* Tools/WebGPUPlayground/CommandLinePlayground/main.c: Added.
(main):
* Tools/WebGPUPlayground/Configurations/AppPlayground.xcconfig: Copied from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/Configurations/Base.xcconfig: Added.
* Tools/WebGPUPlayground/Configurations/CommandLinePlayground.xcconfig: Copied from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/Configurations/DebugRelease.xcconfig: Renamed from Source/WebGPU/Configurations/CommandLinePlayground.xcconfig.
* Tools/WebGPUPlayground/WebGPUPlayground.xcodeproj/project.pbxproj: Added.

Canonical link: https://commits.webkit.org/258998@main
